### PR TITLE
Beautiful Android Notifications

### PR DESCRIPTION
**Pull request by MarcoPolo**
# Preview

## Before
10a02c61-61ca-4505-934e-7dd043e4e17f
* Gross, a new notification for every message!
* "In another conversation" doesn't even look like a different conversation
* Cecile said something! What's the context? Is this in keybase#breaking or keybase#random?
* Who's this `marcopolo` character? Where's his lovely face?

## After
Screen Shot 2019-08-08 at 3 38 00 PM
* Wow, all the notifications are grouped!
* "In Another conversation" is clearly a different conversation. I can see it's in `marcopolo,pkt0`.
* Ah, I can see cecile was responding to my "Testing Avatars in Notifs" message
* Looking good marco!

# Other Niceties

* Device change notifications look better [before](https://user-images.githubusercontent.com/594035/62748199-8f6a6800-ba0c-11e9-8a3d-a58feb4378d3.png) [after](https://user-images.githubusercontent.com/594035/62748218-a9a44600-ba0c-11e9-8da2-93897379fd7a.png)
* Proper support for notification channels on android: 
Screen Shot 2019-08-08 at 6 46 54 PM
* Greatly simplified notification handling on android
* Pass more metadata for chat notifications (Can make iOS notifs nicer too)

@mmaxim for the Go/objc stuff
@jzila for Android stuff

Please let me know if anything looks off, I haven't done much with our Go stack.

## Future work
Reply / Mark Read on Android (I wanted to get this shipped since it's such a big improvement over what we have now)

### CODE REVIEW

File: shared/actions/platform-specific/push.native.tsx [hunk 1]
```diff
@@ -34,30 +34,31 @@ const updateAppBadge = (_, action: NotificationsGen.ReceivedBadgeStatePayload) =
...
-// Push notifications on android are very messy. It works differently if we're entirely killed or if we're in the background
-// The notification is first passed through native code for e.g. plaintext processing.
-// If we were killed, then launching from the notification will go through the
-//   `getStartupDetailsFromInitialPush` flow, via
-//   actions/platfrom-specific/index.native->loadStartupDetails.
-//   * This flow queries the native code's Intent, which contains the original
-//     notification, and then the JS routes us to the right place.
-// If we're backgrounded, then we receive an `androidIntentNotification` event,
-// and execute the listener code below, causing an action that routes us correctly.
+// Push notifications on android are simple.
```
Comment (MarcoPolo) [inline shared/actions/platform-specific/push.native.tsx]: @jzila now they're simple 😛

File: go/bind/keybase.go [hunk 2]
```diff
@@ -51,8 +55,37 @@ var kbfsConfig libkbfs.Config
 var initMutex sync.Mutex
 var initComplete bool
 
+type Person struct {
```
Comment (mmaxim) [inline go/bind/keybase.go]: Let's move all this stuff to a new file in `bind`, call it `notifications.go` or something.

File: go/chat/mobilepush.go [hunk 3]
```diff
@@ -75,79 +73,26 @@ func (h *MobilePush) AckNotificationSuccess(ctx context.Context, pushIDs []strin
...
-				msg.Valid().ClientHeader.TlfName, msg.Valid().MessageBody.Text().Body), nil
-		}
-		return fmt.Sprintf("%s (%s#%s): %s", msg.Valid().SenderUsername,
-			msg.Valid().ClientHeader.TlfName, channelName,
-			msg.Valid().MessageBody.Text().Body), nil
-	default:
-		return fmt.Sprintf("%s: %s", msg.Valid().SenderUsername, msg.Valid().MessageBody.Text().Body), nil
-	}
-}
+func (h *MobilePush) GetConvInfo(ctx context.Context, uid gregor1.UID, convID chat1.ConversationID) (info chat1.ConversationInfoLocal, err error) {
```
Comment (mmaxim) [inline go/chat/mobilepush.go]: Can delete this function, just use `utils.GetVerifiedConv` instead and call it from `bind`

File: go/bind/keybase.go [hunk 4]
```diff
@@ -491,28 +531,73 @@ func HandleBackgroundNotification(strConvID, body string, intMembersType int, di
...
+	// We show avatars on Android
+	if runtime.GOOS == "android" {
+		srvInfo, err := kbSvc.GetHttpSrvInfo()
+		if err == nil {
+			avatar := chat.GetUserAvatar(username, srvInfo)
+			chatNotification.Message.From.KeybaseAvatar = avatar
+		}
 	}
-	age := time.Since(time.Unix(int64(unixTime), 0))
-	if age >= 2*time.Minute {
```
Comment (mmaxim) [inline go/bind/keybase.go]: Restore like this mentioned in chat

File: go/chat/mobilepush.go [hunk 5]
```diff
@@ -75,79 +73,26 @@ func (h *MobilePush) AckNotificationSuccess(ctx context.Context, pushIDs []strin
...
-	case chat1.MessageType_TEXT:
-		return h.formatTextPush(ctx, uid, convID, membersType, msg)
-	case chat1.MessageType_REACTION:
-		return h.formatReactionPush(ctx, uid, convID, membersType, msg)
-	default:
-		h.Debug(ctx, "FormatPushText: unknown message type: %v", msg.GetMessageType())
-		return res, errors.New("invalid message type for plaintext")
+func GetUserAvatar(username string, srvInfo *keybase1.HttpSrvInfo) string {
+	if srvInfo != nil {
+		return fmt.Sprintf("http://%v/av?typ=user&name=%v&format=square_192&token=%v", srvInfo.Address, username, srvInfo.Token)
```
Comment (mmaxim) [inline go/chat/mobilepush.go]: Should make this a function of the `avatars` package, probably by making that `NewSrv` thing get saved somewhere and be accessible. Not a huge deal though, so if it is annoying just skip it.

File: go/bind/notifications.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package keybase
```
Comment (mmaxim) [inline go/bind/notifications.go]: I meant bring over `HandleBackgroundNotification` too

File: go/chat/mobilepush.go [hunk 7]
```diff
@@ -75,79 +72,11 @@ func (h *MobilePush) AckNotificationSuccess(ctx context.Context, pushIDs []strin
...
-	}
-	switch msg.GetMessageType() {
-	case chat1.MessageType_TEXT:
-		return h.formatTextPush(ctx, uid, convID, membersType, msg)
-	case chat1.MessageType_REACTION:
-		return h.formatReactionPush(ctx, uid, convID, membersType, msg)
-	default:
-		h.Debug(ctx, "FormatPushText: unknown message type: %v", msg.GetMessageType())
-		return res, errors.New("invalid message type for plaintext")
+func GetUserAvatar(username string, srvInfo *keybase1.HttpSrvInfo) string {
```
Comment (mmaxim) [inline go/chat/mobilepush.go]: How come this is still here? I see you have it in avatar land too

File: shared/android/app/src/main/java/io/keybase/ossifrage/KBReactPackage.java [hunk 8]
```diff
@@ -11,7 +11,6 @@
 import java.util.List;
 
 import io.keybase.ossifrage.components.KBTextInputManager;
-import io.keybase.ossifrage.modules.IntentHandler;
```
Comment (jzila) [inline shared/android/app/src/main/java/io/keybase/ossifrage/KBReactPackage.java]: So this is now done by the react module?

File: shared/android/app/src/main/java/io/keybase/ossifrage/KeybasePushNotificationListenerService.java [hunk 9]
```diff
@@ -16,17 +22,52 @@
...
 
-public class KeybasePushNotificationListenerService extends RNPushNotificationListenerService {
+public class KeybasePushNotificationListenerService extends FirebaseMessagingService {
+    public static String CHAT_CHANNEL_ID = "kb_chat_channel";
+    public static String FOLLOW_CHANNEL_ID = "kb_follow_channel";
+    public static String DEVICE_CHANNEL_ID = "kb_device_channel";
+    public static String GENERAL_CHANNEL_ID = "kb_rest_channel";
+
+    // This keeps a small (5 msg) ring buffer cache of the last messages the user was notified about,
+    // so we can give context to future notifications
```
Comment (jzila) [inline shared/android/app/src/main/java/io/keybase/ossifrage/KeybasePushNotificationListenerService.java]: 5 msg per conversation

File: go/chat/mobilepush.go (hunk 7 above)
Comment (MarcoPolo) [reply go/chat/mobilepush.go]: I forgot to remove this!

File: shared/actions/platform-specific/index.native.tsx [hunk 10]
```diff
@@ -207,28 +206,6 @@ const updateChangedFocus = (_: Container.TypedState, action: ConfigGen.MobileApp
   return ConfigGen.createChangedFocus({appFocused})
 }
 
-const getStartupDetailsFromShare = (): Promise<
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/actions/platform-specific/index.native.tsx]: Logic was deduped. We now only handle this here: https://github.com/keybase/client/pull/18852/files#diff-270c1a26c4443ac1cfc4c6a7b223b13fR66

[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/actions/platform-specific/index.native.tsx]: https://github.com/keybase/client/blob/57a27b6915cde8020f8728c8f03d272eee4001f2/shared/actions/platform-specific/push.native.tsx#L66

File: shared/android/app/src/main/java/io/keybase/ossifrage/KBPushNotifier.java [hunk 11]
```diff
@@ -1,56 +1,219 @@
...
+
+
+    NotificationCompat.Builder builder =
+      new NotificationCompat.Builder(this.context, KeybasePushNotificationListenerService.CHAT_CHANNEL_ID)
+        .setSmallIcon(R.drawable.ic_notif)
+        .setContentIntent(pending_intent)
+        .setAutoCancel(true);
+
+    Message msg = chatNotification.getMessage();
+    Person.Builder personBuilder = new Person.Builder().setName(msg.getFrom().getKeybaseUsername()).setBot(msg.getFrom().getIsBot());
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/android/app/src/main/java/io/keybase/ossifrage/KBPushNotifier.java]: We need to fix the formatting of the java code in general... 😅

File: shared/android/app/src/main/java/io/keybase/ossifrage/KeybasePushNotificationListenerService.java [hunk 12]
```diff
@@ -73,30 +170,35 @@ public void onMessageReceived(RemoteMessage message) {
...
+                    int membersType = Integer.parseInt(bundle.getString("t"));
                     String convID = bundle.getString("c");
-                    Integer messageId = Integer.parseInt(bundle.getString("d"));
+                    int messageId = Integer.parseInt(bundle.getString("d"));
                     JSONArray pushes = parseJSONArray(bundle.getString("p"));
                     String pushId = pushes.getString(0);
-                    Integer badgeCount = Integer.parseInt(bundle.getString("b"));
-                    Integer unixTime = Integer.parseInt(bundle.getString("x"));
+                    int badgeCount = Integer.parseInt(bundle.getString("b"));
+                    int unixTime = Integer.parseInt(bundle.getString("x"));
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/android/app/src/main/java/io/keybase/ossifrage/KeybasePushNotificationListenerService.java]: Android Studio's lint

File: shared/android/app/src/main/java/io/keybase/ossifrage/KBReactPackage.java (hunk 8 above)
Comment (MarcoPolo) [reply shared/android/app/src/main/java/io/keybase/ossifrage/KBReactPackage.java]: You mean the intent data passing? That happens in main activity. See this comment: https://github.com/keybase/client/pull/18852/files#diff-270c1a26c4443ac1cfc4c6a7b223b13fR37

File: shared/android/app/src/main/java/io/keybase/ossifrage/KeybasePushNotificationListenerService.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/android/app/src/main/java/io/keybase/ossifrage/KeybasePushNotificationListenerService.java]: We were boxing them originally. Integer.parseInt returns an int. https://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#parseInt(java.lang.String)

File: shared/android/app/src/main/java/io/keybase/ossifrage/KBPushNotifier.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/android/app/src/main/java/io/keybase/ossifrage/KBPushNotifier.java]: https://github.com/keybase/client/compare/marco/better-android-notifs...marco/fix-java-styling?expand=1 (after this pr)

File: go/bind/notifications.go [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+		return err
+	}
+
+	conv, err := utils.GetVerifiedConv(ctx, gc, uid, convID, types.InboxSourceDataSourceAll)
+	if err != nil {
+		kbCtx.Log.CDebugf(ctx, "Failed to get conversation info", err)
+		return err
+	}
+
+	// TODO how to figure out if it's a bot?
```
Comment (joshblum) [inline go/bind/notifications.go]: `isBot := gregor1.UITPtrEq(msg.Valid().ClientHeader.BotUID, msg.Valid().ClientHeader.Sender)`

you can put this as a helper in `chat1/extras.go` if it will be useful elsewhere..

File: go/bind/notifications.go [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+	username := msgUnboxed.Valid().SenderUsername
+
+	chatNotification := ChatNotification{
+		IsPlaintext: false,
```
Comment (joshblum) [inline go/bind/notifications.go]: why not `displayPlaintext`?

**Comment by MarcoPolo** [comment]
Manual Testing Strategy:

| iOS | Android | Test |
|-----|---------|-----|
| [ ] |  [ ] | 1 on 1 message - plaintext |
| [ ] |  [ ] | 1 on 1 message - not plain |
| [ ] |  [ ] | Exploding Message |
| [ ] |  [ ] | Big Team pushes with Channel Name |
| [ ] |  [ ] | Small Team |

File: go/bind/notifications.go [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+		avatar, err := kbSvc.GetUserAvatar(username)
+
+		if err != nil {
+			kbCtx.Log.CDebugf(ctx, "Push Notif: Err in getting user avatar %v", err)
+		} else {
+			chatNotification.Message.From.KeybaseAvatar = avatar
+		}
+	}
+
+	chatNotification.IsPlaintext = true
```
Comment (joshblum) [inline go/bind/notifications.go]: can rm now

**Comment by MarcoPolo** [comment]
I fixed exploding messages since I didn't realize we only get those in chat.newmessage. I think we should fix up that flow because it will basically always error out of `handleBackgroundNotification` on ios with a stale notification because the unix time is 0. It would be nice to get more meta data too so we can show that on the android side (instead of falling back to the unstyled generic type message)

File: shared/android/app/src/main/java/io/keybase/ossifrage/KBPushNotifier.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jzila) [reply shared/android/app/src/main/java/io/keybase/ossifrage/KBPushNotifier.java]: Cool, what command did you run to generate those formatting fixes?

File: shared/android/app/src/main/java/io/keybase/ossifrage/KBPushNotifier.java [hunk 16]
```diff
@@ -1,56 +1,214 @@
...
+    PendingIntent pending_intent = PendingIntent.getActivity(this.context, 0, open_activity_intent,
+      PendingIntent.FLAG_UPDATE_CURRENT);
+
+    return pending_intent;
+  }
+
+  private IconCompat getKeybaseAvatar(String avatarUri) {
+    IconCompat icon = null;
+
+    if (!avatarUri.isEmpty()) {
```
Comment (jzila) [inline shared/android/app/src/main/java/io/keybase/ossifrage/KBPushNotifier.java]: To reduce nesting, I'd prefer:
```suggestion
    if (avatarUri.isEmpty()) return null;
    ...
```
Then you don't even need the `icon` declaration above, instead returning directly where it's needed.

File: shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java [hunk 17]
```diff
@@ -37,142 +49,282 @@
...
+    return filePath;
+  }
 
-    @Override
-    protected void onResume() {
-        super.onResume();
-        Keybase.setAppStateForeground();
-    }
+  // Emits the intent to JS
+  private void emitIntent(Intent intent) {
```
Comment (jzila) [inline shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java]: Why was this moved into `MainActivity` instead of being a separate class as before? I think this makes it harder to read and understand what's going on. This method doesn't use anything in this class, so I don't see why it can't be separated out.

File: shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java [hunk 18]
```diff
@@ -37,142 +49,282 @@
...
-    @Override
-    protected String getMainComponentName() {
-        return "Keybase";
+    // We need to run this on the main thread, as the React code assumes that is true.
+    // Namely, DevServerHelper constructs a Handler() without a Looper, which triggers:
+    // "Can't create handler inside thread that has not called Looper.prepare()"
+    Handler handler = new Handler(Looper.getMainLooper());
+    handler.post(() -> {
+      // Construct and load our normal React JS code bundle
+      ReactInstanceManager mReactInstanceManager = ((ReactApplication) getApplication()).getReactNativeHost().getReactInstanceManager();
```
Comment (jzila) [inline shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java]: Why is this called `mReactInstanceManager`? It's not a member variable, it's a local variable...

Comment (MarcoPolo) [reply shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java]: copy/pasted, but good point :+1:

File: shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java (hunk 17 above)
Comment (MarcoPolo) [reply shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java]: Because this is only used by this class. That's why it's private. I can move it to it's own class but you don't get the same guarantees.

**Comment by jzila** [comment]
Very excited to use this!
